tests/utils_tests/test_choices.py

```diff
@@ -0,0 +1,216 @@
+        choices = [
+            ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]],
+            ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]],
+            ["unknown", _("Unknown")],
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_iterator_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        def generator():
+            yield ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]]
+            yield ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]]
+            yield ["unknown", _("Unknown")]
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_mixed_mapping_and_iterable(self):
+        # Although not documented as it's better to stick to either mappings or
+        # iterables, ensure that this works as expected.
+        choices = {
+            "Audio": [("vinyl", _("Vinyl")), ("cd", _("CD"))],
+            "Video": [("vhs", _("VHS Tape")), ("dvd", _("DVD"))],
+            "unknown": _("Unknown"),
+        }
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+        choices = [
+            ("Audio", {"vinyl": _("Vinyl"), "cd": _("CD")}),
+            ("Video", {"vhs": _("VHS Tape"), "dvd": _("DVD")}),
+            ("unknown", _("Unknown")),
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_iterable_set(self):
+        # Although not documented as sets are unordered which results in randomised
+        # order in form fields, ensure that this works as expected. Note that consistent
+        # ordering in migrations is enforced by the migrations serializer.
+        choices = {
+            ("C", _("Club")),
+            ("D", _("Diamond")),
+            ("H", _("Heart")),
+            ("S", _("Spade")),
+        }
+        self.assertEqual(
+            sorted(normalize_field_choices(choices)),
+            sorted(self.expected),
+        )
+
+    def test_unsupported_values_returned_unmodified(self):
+        # Unsupported values must be returned unmodified for model system check to work.
+        unsupported_values = [
+            1j,
+            123,
+            123.45,
+            "invalid",
+            b"invalid",
+            _("invalid"),
+            object(),
+            None,
+            True,
+            False,
+            # Special cases of a string-likes which would unpack incorrectly.
+            ["ab"],
+            [b"ab"],
+            [_("ab")],
+            # Non-iterable items or iterable items with incorrect number of elements.
+            [123],  # TypeError: cannot unpack non-iterable int object
+            [("value",)],  # ValueError: not enough values to unpack
+            [("value", "label", "other")],  # ValueError: too many values to unpack
+        ]
```

Comment:
So I [added a separate test for the empty cases](https://github.com/django/django/pull/16943/commits/cbd9bf9ea839d432835842abdcdd6d0c944f48fb#diff-d485e2a2d7b8491e3211e00184bc104bcc6b19ff0d8162fe9b7ad5cc9cb826abR24-R30) because it's expected that we'll get these valid empty iterables and they should be normalized to `[]`. This particular test is for abnormal values that we want to return unmodified so that they are picked up by the system check... _(Hence my [comment](https://github.com/django/django/pull/16943#discussion_r1270597201) above about whether we could simplify the system check stuff somehow rather than iterating through it all over again...)_